adapters/src/main/java/com/squareup/moshi/adapters/RuntimeJsonAdapterFactory.java

```diff
@@ -63,112 +75,117 @@
 
-    RuntimeJsonAdapter(String labelKey, Map<String, JsonAdapter<Object>> labelToAdapter,
-        Map<Type, String> typeToLabel, JsonAdapter<Object> objectJsonAdapter) {
+    /** Single-element options containing the label's key only. */
+    final JsonReader.Options labelKeyOptions;
+    /** Corresponds to subtypes. */
+    final JsonReader.Options labelOptions;
+
+    RuntimeJsonAdapter(String labelKey, List<String> labels,
+        List<Type> subtypes, List<JsonAdapter<Object>> jsonAdapters,
+        JsonAdapter<Object> objectJsonAdapter) {
       this.labelKey = labelKey;
-      this.labelToAdapter = labelToAdapter;
-      this.typeToLabel = typeToLabel;
+      this.labels = labels;
+      this.subtypes = subtypes;
+      this.jsonAdapters = jsonAdapters;
       this.objectJsonAdapter = objectJsonAdapter;
+
+      this.labelKeyOptions = JsonReader.Options.of(labelKey);
+      this.labelOptions = JsonReader.Options.of(labels.toArray(new String[0]));
     }
 
     @Override public Object fromJson(JsonReader reader) throws IOException {
-      JsonReader.Token peekedToken = reader.peek();
-      if (peekedToken != JsonReader.Token.BEGIN_OBJECT) {
-        throw new JsonDataException("Expected BEGIN_OBJECT but was " + peekedToken
-            + " at path " + reader.getPath());
-      }
-      Object jsonValue = reader.readJsonValue();
-      Map<String, Object> jsonObject = (Map<String, Object>) jsonValue;
-      Object label = jsonObject.get(labelKey);
-      if (label == null) {
-        throw new JsonDataException("Missing label for " + labelKey);
-      }
-      if (!(label instanceof String)) {
-        throw new JsonDataException("Label for '"
-            + labelKey
-            + "' must be a string but was "
-            + label
-            + ", a "
-            + label.getClass());
-      }
-      JsonAdapter<Object> adapter = labelToAdapter.get(label);
-      if (adapter == null) {
-        throw new JsonDataException("Expected one of "
-            + labelToAdapter.keySet()
-            + " for key '"
-            + labelKey
-            + "' but found '"
-            + label
-            + "'. Register a subtype for this label.");
+      int labelIndex = labelIndex(reader.peekJson());
+      return jsonAdapters.get(labelIndex).fromJson(reader);
+    }
+
+    private int labelIndex(JsonReader reader) throws IOException {
+      reader.beginObject();
+      while (reader.hasNext()) {
+        if (reader.selectName(labelKeyOptions) == -1) {
+          reader.skipName();
+          reader.skipValue();
+          continue;
+        }
+
+        int labelIndex = reader.selectString(labelOptions);
+        if (labelIndex == -1) {
+          throw new JsonDataException("Expected one of "
+              + labels
+              + " for key '"
+              + labelKey
+              + "' but found '"
+              + reader.nextString()
+              + "'. Register a subtype for this label.");
+        }
+        reader.close();
```

Comment:
not really necessary, right?

Reply:
I wanna fix PeekingSource so that it clears buffers on close. That could potentially allow us to save some GC load.

Reply:
cool 👍